Duplicates found in MROs (OrderedSet)

### Steps to reproduce
This issue is directly linked to PyCQA/pylint#4093

1. Write the following module:
```
from ordered_set import OrderedSet

class RegistryOrderedSet:
    def __init__(self, clazz, type_name="Object"):
        self._objects_dict = defaultdict(OrderedSet)
```
2. Lint it

### Current behavior
```
Traceback (most recent call last):
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/decorators.py", line 34, in cached
    return cache[func]
KeyError: <bound method ClassDef.slots of <ClassDef.OrderedSet l.55 at 0x7f2517b98dc0>>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/peillexg/PROGRAMMATION/Python/pylint_sandbox/bug_pylint_4093/test/bin/pylint", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/peillexg/PROGRAMMATION/Python/pylint/bin/pylint", line 5, in <module>
    run_pylint()
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/__init__.py", line 22, in run_pylint
    PylintRun(sys.argv[1:])
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/run.py", line 358, in __init__
    linter.check(args)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 862, in check
    self._check_files(
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 896, in _check_files
    self._check_file(get_ast, check_astroid_module, name, filepath, modname)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 922, in _check_file
    check_astroid_module(ast_node)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 1054, in check_astroid_module
    retval = self._check_astroid_module(
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/lint/pylinter.py", line 1099, in _check_astroid_module
    walker.walk(ast_node)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/utils/ast_walker.py", line 75, in walk
    self.walk(child)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/utils/ast_walker.py", line 72, in walk
    callback(astroid)
  File "/home/peillexg/PROGRAMMATION/Python/pylint/pylint/checkers/variables.py", line 1182, in visit_importfrom
    module = node.do_import_module(name_parts[0])
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/mixins.py", line 99, in do_import_module
    return mymodule.import_module(
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 646, in import_module
    return MANAGER.ast_from_module_name(absmodname)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/manager.py", line 191, in ast_from_module_name
    return self.ast_from_file(found_spec.location, modname, fallback=False)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/manager.py", line 100, in ast_from_file
    return AstroidBuilder(self).file_build(filepath, modname)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/builder.py", line 139, in file_build
    return self._post_build(module, encoding)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/builder.py", line 159, in _post_build
    self.delayed_assattr(delayed)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/builder.py", line 235, in delayed_assattr
    if not _can_assign_attr(inferred, node.attrname):
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/builder.py", line 60, in _can_assign_attr
    slots = node.slots()
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/decorators.py", line 36, in cached
    cache[func] = result = func(*args, **kwargs)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 2838, in slots
    slots = list(grouped_slots())
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 2823, in grouped_slots
    for cls in self.mro()[:-1]:
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 2909, in mro
    return self._compute_mro(context=context)
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 2898, in _compute_mro
    unmerged_mro = list(clean_duplicates_mro(unmerged_mro, self, context))
  File "/home/peillexg/PROGRAMMATION/Python/astroid/astroid/scoped_nodes.py", line 110, in clean_duplicates_mro
    raise exceptions.DuplicateBasesError(
astroid.exceptions.DuplicateBasesError: Duplicates found in MROs (OrderedSet), (_GenericAlias, _Final, object), (_GenericAlias, _Final, object), (_GenericAlias, _GenericAlias) for <ClassDef.OrderedSet l.55 at 0x7f2517b98dc0>
```

### Expected behavior
No error. It was the case with `astroid 2.4.2`. A bisection leads to commit 7f1d5134d89aab9b3e49030676637debfd731dbb.

### ``python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"`` output
2.5

Found a similar issue with `aiohttp`.
```
pylint 2.7.0
astroid 1.5.0
```

Interestingly the error occurred when using Python **3.8**. However, 3.9 was fine.

```py
# test_pylint.py

# pylint: disable=missing-class-docstring,inherit-non-class,too-few-public-methods
# pylint: disable=no-value-for-parameter,unused-import,missing-module-docstring
# pylint: disable=super-init-not-called
import abc
from typing import Sized, Iterable


class AbstractRoute(abc.ABC):
    pass


class AbstractResource(Sized, Iterable["AbstractRoute"]):
    pass


class IndexView(AbstractResource):
    def __init__(self):
        self.var = 1
```

Running `pylint test.py` results in
```
Traceback (most recent call last):
  File ".../astroid/astroid/decorators.py", line 34, in cached
    return cache[func]
KeyError: <bound method ClassDef.slots of <ClassDef.IndexView l.16 at 0x7fdfae20e460>>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../pylint/pylint/lint/pylinter.py", line 1030, in get_ast
    return MANAGER.ast_from_file(filepath, modname, source=True)
  File ".../astroid/astroid/manager.py", line 100, in ast_from_file
    return AstroidBuilder(self).file_build(filepath, modname)
  File ".../astroid/astroid/builder.py", line 138, in file_build
    return self._post_build(module, encoding)
  File ".../astroid/astroid/builder.py", line 158, in _post_build
    self.delayed_assattr(delayed)
  File ".../astroid/astroid/builder.py", line 234, in delayed_assattr
    if not _can_assign_attr(inferred, node.attrname):
  File ".../astroid/astroid/builder.py", line 59, in _can_assign_attr
    slots = node.slots()
  File ".../astroid/astroid/decorators.py", line 36, in cached
    cache[func] = result = func(*args, **kwargs)
  File ".../astroid/astroid/scoped_nodes.py", line 2834, in slots
    slots = list(grouped_slots())
  File ".../astroid/astroid/scoped_nodes.py", line 2819, in grouped_slots
    for cls in self.mro()[:-1]:
  File ".../astroid/astroid/scoped_nodes.py", line 2905, in mro
    return self._compute_mro(context=context)
  File ".../astroid/astroid/scoped_nodes.py", line 2882, in _compute_mro
    mro = base._compute_mro(context=context)
  File ".../astroid/astroid/scoped_nodes.py", line 2894, in _compute_mro
    unmerged_mro = list(clean_duplicates_mro(unmerged_mro, self, context))
  File ".../astroid/astroid/scoped_nodes.py", line 109, in clean_duplicates_mro
    raise exceptions.DuplicateBasesError(
astroid.exceptions.DuplicateBasesError: Duplicates found in MROs (AbstractResource), (_GenericAlias, _Final, object), (_GenericAlias, _Final, object), (_GenericAlias, _GenericAlias) for <ClassDef.AbstractResource l.12 at 0x7fdfae20e250>.
************* Module test_pylint
test_pylint.py:1:0: F0002: <class 'astroid.exceptions.DuplicateBasesError'>: Duplicates found in MROs (AbstractResource), (_GenericAlias, _Final, object), (_GenericAlias, _Final, object), (_GenericAlias, _GenericAlias) for <ClassDef.AbstractResource l.12 at 0x7fdfae20e250>. (astroid-error)
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -15,6 +15,14 @@ Release Date: 2021-02-28
 
 * Improve typing.TypedDict inference
 
+* Fix the `Duplicates found in MROs` false positive.
+
+  Closes #905
+  Closes PyCQA/pylint#4093
+  Closes PyCQA/pylint#4131
+  Closes PyCQA/pylint#4145
+  Closes PyCQA/pylint#3247
+
 
 What's New in astroid 2.5?
 ============================

--- a/ChangeLog
+++ b/ChangeLog
@@ -18,10 +18,11 @@ Release Date: 2021-02-28
 * Fix the `Duplicates found in MROs` false positive.
 
   Closes #905
+  Closes PyCQA/pylint#2717
+  Closes PyCQA/pylint#3247
   Closes PyCQA/pylint#4093
   Closes PyCQA/pylint#4131
   Closes PyCQA/pylint#4145
-  Closes PyCQA/pylint#3247
 
 
 What's New in astroid 2.5?

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -132,10 +132,10 @@ def _looks_like_typing_alias(node: nodes.Call) -> bool:
     :param node: call node
     """
     return (
-        isinstance(node, nodes.Call) and
-        isinstance(node.func, nodes.Name) and
-        node.func.name == "_alias" and
-        isinstance(node.args[0], nodes.Attribute)
+        isinstance(node, nodes.Call)
+        and isinstance(node.func, nodes.Name)
+        and node.func.name == "_alias"
+        and isinstance(node.args[0], nodes.Attribute)
     )
 
 

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -14,10 +14,12 @@ from astroid import (
     UseInferenceDefault,
     extract_node,
     inference_tip,
+    node_classes,
     nodes,
     context,
     InferenceError,
 )
+import astroid
 
 PY39 = sys.version_info[:2] >= (3, 9)
 
@@ -89,6 +91,7 @@ def infer_typing_attr(node, context=None):
     node = extract_node(TYPING_TYPE_TEMPLATE.format(value.qname().split(".")[-1]))
     return node.infer(context=context)
 
+
 GET_ITEM_TEMPLATE = """
 @classmethod
 def __getitem__(cls, value):
@@ -118,18 +121,33 @@ def infer_typedDict(  # pylint: disable=invalid-name
     node.root().locals["TypedDict"] = [class_def]
 
 
-def _looks_like_typing_alias(node):
+def _looks_like_typing_alias(node: nodes.Call) -> bool:
+    """
+    Returns True if the node corresponds to a call to _alias function.
+    For example :
+
+    MutableSet = _alias(collections.abc.MutableSet, T)
+
+    :param node: call node
+    """
     if isinstance(node, nodes.Call) and isinstance(node.func, nodes.Name):
-        if node.func.name == "_alias":
-            if isinstance(node.args[0], nodes.Attribute):
-                return True
+        if node.func.name == "_alias" and isinstance(node.args[0], nodes.Attribute):
+            return True
     return False
 
 
-def infer_typing_alias(node, context=None):
+def infer_typing_alias(node: nodes.Call, context: context.InferenceContext = None) -> node_classes.NodeNG:
+    """
+    Infers the call to _alias function
+
+    :param node: call node
+    :param context: inference context
+    """
     if not isinstance(node, nodes.Call):
         return
     res = next(node.args[0].infer(context=context))
+    # Needs to mock the __getitem__ class method so that
+    # MutableSet[T] is acceptable
     func_to_add = extract_node(GET_ITEM_TEMPLATE)
     if res.metaclass():
         res.metaclass().locals['__getitem__'] = [func_to_add]

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -89,6 +89,12 @@ def infer_typing_attr(node, context=None):
     node = extract_node(TYPING_TYPE_TEMPLATE.format(value.qname().split(".")[-1]))
     return node.infer(context=context)
 
+GET_ITEM_TEMPLATE = """
+@classmethod
+def __getitem__(cls, value):
+    return cls
+"""
+
 
 def _looks_like_typedDict(  # pylint: disable=invalid-name
     node: nodes.FunctionDef,
@@ -123,9 +129,10 @@ def _looks_like_typing_alias(node):
 def infer_typing_alias(node, context=None):
     if not isinstance(node, nodes.Call):
         return
-    node = extract_node(TYPING_TYPE_TEMPLATE.format(node.args[0].attrname))
-    res = next(node.infer(context=context))
-    # breakpoint()
+    res = next(node.args[0].infer(context=context))
+    func_to_add = extract_node(GET_ITEM_TEMPLATE)
+    if res.metaclass():
+        res.metaclass().locals['__getitem__'] = [func_to_add]
     return res
 
 

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -180,13 +180,13 @@ def infer_typing_alias(
 
     if res != astroid.Uninferable and isinstance(res, nodes.ClassDef):
         class_def = nodes.ClassDef(
-            name=res.name,
-            lineno=res.lineno,
-            col_offset=res.col_offset,
+            name=f"{res.name}_typing",
+            lineno=0,
+            col_offset=0,
             parent=res.parent,
         )
         class_def.postinit(
-            bases=res.bases,
+            bases=[res],
             body=res.body,
             decorators=res.decorators,
             metaclass=create_typing_metaclass(),
@@ -196,8 +196,8 @@ def infer_typing_alias(
     if len(node.args) == 2 and isinstance(node.args[0], nodes.Attribute):
         class_def = nodes.ClassDef(
             name=node.args[0].attrname,
-            lineno=node.lineno,
-            col_offset=node.col_offset,
+            lineno=0,
+            col_offset=0,
             parent=node.parent,
         )
         class_def.postinit(

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -112,6 +112,23 @@ def infer_typedDict(  # pylint: disable=invalid-name
     node.root().locals["TypedDict"] = [class_def]
 
 
+def _looks_like_typing_alias(node):
+    if isinstance(node, nodes.Call) and isinstance(node.func, nodes.Name):
+        if node.func.name == "_alias":
+            if isinstance(node.args[0], nodes.Attribute):
+                return True
+    return False
+
+
+def infer_typing_alias(node, context=None):
+    if not isinstance(node, nodes.Call):
+        return
+    node = extract_node(TYPING_TYPE_TEMPLATE.format(node.args[0].attrname))
+    res = next(node.infer(context=context))
+    # breakpoint()
+    return res
+
+
 MANAGER.register_transform(
     nodes.Call,
     inference_tip(infer_typing_typevar_or_newtype),
@@ -125,3 +142,7 @@ if PY39:
     MANAGER.register_transform(
         nodes.FunctionDef, infer_typedDict, _looks_like_typedDict
     )
+
+MANAGER.register_transform(
+    nodes.Call, infer_typing_alias, _looks_like_typing_alias
+)

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -94,6 +94,28 @@ def infer_typing_attr(node, context=None):
     return node.infer(context=context)
 
 
+def _looks_like_typedDict(  # pylint: disable=invalid-name
+    node: nodes.FunctionDef,
+) -> bool:
+    """Check if node is TypedDict FunctionDef."""
+    return isinstance(node, nodes.FunctionDef) and node.name == "TypedDict"
+
+
+def infer_typedDict(  # pylint: disable=invalid-name
+    node: nodes.FunctionDef, ctx: context.InferenceContext = None
+) -> None:
+    """Replace TypedDict FunctionDef with ClassDef."""
+    class_def = nodes.ClassDef(
+        name="TypedDict",
+        doc=node.doc,
+        lineno=node.lineno,
+        col_offset=node.col_offset,
+        parent=node.parent,
+    )
+    class_def.postinit(bases=[], body=[], decorators=None)
+    node.root().locals["TypedDict"] = [class_def]
+
+
 GET_ITEM_TEMPLATE = """
 @classmethod
 def __getitem__(cls, value):
@@ -124,28 +146,6 @@ def create_typing_metaclass():
     )
     typing_meta.locals["__getitem__"] = [func_to_add]
     return typing_meta
-
-
-def _looks_like_typedDict(  # pylint: disable=invalid-name
-    node: nodes.FunctionDef,
-) -> bool:
-    """Check if node is TypedDict FunctionDef."""
-    return isinstance(node, nodes.FunctionDef) and node.name == "TypedDict"
-
-
-def infer_typedDict(  # pylint: disable=invalid-name
-    node: nodes.FunctionDef, ctx: context.InferenceContext = None
-) -> None:
-    """Replace TypedDict FunctionDef with ClassDef."""
-    class_def = nodes.ClassDef(
-        name="TypedDict",
-        doc=node.doc,
-        lineno=node.lineno,
-        col_offset=node.col_offset,
-        parent=node.parent,
-    )
-    class_def.postinit(bases=[], body=[], decorators=None)
-    node.root().locals["TypedDict"] = [class_def]
 
 
 def _looks_like_typing_alias(node: nodes.Call) -> bool:

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -8,6 +8,7 @@
 """Astroid hooks for typing.py support."""
 import sys
 import typing
+from functools import lru_cache
 
 from astroid import (
     MANAGER,
@@ -99,6 +100,31 @@ def __getitem__(cls, value):
     return cls
 """
 
+ABC_METACLASS_TEMPLATE = """
+from abc import ABCMeta
+ABCMeta
+"""
+
+
+@lru_cache()
+def create_typing_metaclass():
+    #  Needs to mock the __getitem__ class method so that
+    #  MutableSet[T] is acceptable
+    func_to_add = extract_node(GET_ITEM_TEMPLATE)
+
+    abc_meta = next(extract_node(ABC_METACLASS_TEMPLATE).infer())
+    typing_meta = nodes.ClassDef(
+        name="ABCMeta_typing",
+        lineno=abc_meta.lineno,
+        col_offset=abc_meta.col_offset,
+        parent=abc_meta.parent,
+    )
+    typing_meta.postinit(
+        bases=[extract_node(ABC_METACLASS_TEMPLATE)], body=[], decorators=None
+    )
+    typing_meta.locals["__getitem__"] = [func_to_add]
+    return typing_meta
+
 
 def _looks_like_typedDict(  # pylint: disable=invalid-name
     node: nodes.FunctionDef,
@@ -140,8 +166,8 @@ def _looks_like_typing_alias(node: nodes.Call) -> bool:
 
 
 def infer_typing_alias(
-    node: nodes.Call, context: context.InferenceContext = None
-) -> node_classes.NodeNG:
+    node: nodes.Call, ctx: context.InferenceContext = None
+) -> typing.Optional[node_classes.NodeNG]:
     """
     Infers the call to _alias function
 
@@ -149,14 +175,37 @@ def infer_typing_alias(
     :param context: inference context
     """
     if not isinstance(node, nodes.Call):
-        return
-    res = next(node.args[0].infer(context=context))
-    #  Needs to mock the __getitem__ class method so that
-    #  MutableSet[T] is acceptable
-    func_to_add = extract_node(GET_ITEM_TEMPLATE)
-    if res.metaclass():
-        res.metaclass().locals["__getitem__"] = [func_to_add]
-    return res
+        return None
+    res = next(node.args[0].infer(context=ctx))
+
+    if res != astroid.Uninferable and isinstance(res, nodes.ClassDef):
+        class_def = nodes.ClassDef(
+            name=res.name,
+            lineno=res.lineno,
+            col_offset=res.col_offset,
+            parent=res.parent,
+        )
+        class_def.postinit(
+            bases=res.bases,
+            body=res.body,
+            decorators=res.decorators,
+            metaclass=create_typing_metaclass(),
+        )
+        return class_def
+
+    if len(node.args) == 2 and isinstance(node.args[0], nodes.Attribute):
+        class_def = nodes.ClassDef(
+            name=node.args[0].attrname,
+            lineno=node.lineno,
+            col_offset=node.col_offset,
+            parent=node.parent,
+        )
+        class_def.postinit(
+            bases=[], body=[], decorators=None, metaclass=create_typing_metaclass()
+        )
+        return class_def
+
+    return None
 
 
 MANAGER.register_transform(

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -136,7 +136,9 @@ def _looks_like_typing_alias(node: nodes.Call) -> bool:
     return False
 
 
-def infer_typing_alias(node: nodes.Call, context: context.InferenceContext = None) -> node_classes.NodeNG:
+def infer_typing_alias(
+    node: nodes.Call, context: context.InferenceContext = None
+) -> node_classes.NodeNG:
     """
     Infers the call to _alias function
 
@@ -146,11 +148,11 @@ def infer_typing_alias(node: nodes.Call, context: context.InferenceContext = Non
     if not isinstance(node, nodes.Call):
         return
     res = next(node.args[0].infer(context=context))
-    # Needs to mock the __getitem__ class method so that
-    # MutableSet[T] is acceptable
+    #  Needs to mock the __getitem__ class method so that
+    #  MutableSet[T] is acceptable
     func_to_add = extract_node(GET_ITEM_TEMPLATE)
     if res.metaclass():
-        res.metaclass().locals['__getitem__'] = [func_to_add]
+        res.metaclass().locals["__getitem__"] = [func_to_add]
     return res
 
 
@@ -168,6 +170,4 @@ if PY39:
         nodes.FunctionDef, infer_typedDict, _looks_like_typedDict
     )
 
-MANAGER.register_transform(
-    nodes.Call, infer_typing_alias, _looks_like_typing_alias
-)
+MANAGER.register_transform(nodes.Call, infer_typing_alias, _looks_like_typing_alias)

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -21,6 +21,7 @@ from astroid import (
 )
 import astroid
 
+PY37 = sys.version_info[:2] >= (3, 7)
 PY39 = sys.version_info[:2] >= (3, 9)
 
 TYPING_NAMEDTUPLE_BASENAMES = {"NamedTuple", "typing.NamedTuple"}
@@ -131,9 +132,9 @@ def _looks_like_typing_alias(node: nodes.Call) -> bool:
     :param node: call node
     """
     return (
-        isinstance(node, nodes.Call) and 
+        isinstance(node, nodes.Call) and
         isinstance(node.func, nodes.Name) and
-        node.func.name == "_alias" and 
+        node.func.name == "_alias" and
         isinstance(node.args[0], nodes.Attribute)
     )
 
@@ -172,4 +173,5 @@ if PY39:
         nodes.FunctionDef, infer_typedDict, _looks_like_typedDict
     )
 
-MANAGER.register_transform(nodes.Call, infer_typing_alias, _looks_like_typing_alias)
+if PY37:
+    MANAGER.register_transform(nodes.Call, infer_typing_alias, _looks_like_typing_alias)

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -130,10 +130,12 @@ def _looks_like_typing_alias(node: nodes.Call) -> bool:
 
     :param node: call node
     """
-    if isinstance(node, nodes.Call) and isinstance(node.func, nodes.Name):
-        if node.func.name == "_alias" and isinstance(node.args[0], nodes.Attribute):
-            return True
-    return False
+    return (
+        isinstance(node, nodes.Call) and 
+        isinstance(node.func, nodes.Name) and
+        node.func.name == "_alias" and 
+        isinstance(node.args[0], nodes.Attribute)
+    )
 
 
 def infer_typing_alias(

--- a/tests/unittest_brain.py
+++ b/tests/unittest_brain.py
@@ -1213,7 +1213,7 @@ class TypingBrain(unittest.TestCase):
         correctly inferred.
         """
         node = builder.extract_node(
-        """
+            """
         from typing import TypeVar, MutableSet
 
         T = TypeVar("T")

--- a/tests/unittest_brain.py
+++ b/tests/unittest_brain.py
@@ -36,6 +36,7 @@ import io
 import queue
 import re
 import os
+from typing import Collection, MutableSet
 
 try:
     import multiprocessing  # pylint: disable=unused-import
@@ -1205,6 +1206,41 @@ class TypingBrain(unittest.TestCase):
         typing_module = inferred_base.root()
         assert len(typing_module.locals["TypedDict"]) == 1
         assert inferred_base == typing_module.locals["TypedDict"][0]
+
+    def test_typing_alias_type(self):
+        """
+        Test that the type aliased thanks to typing._alias function are
+        correctly inferred.
+        """
+        node = builder.extract_node(
+        """
+        from typing import TypeVar, MutableSet
+
+        T = TypeVar("T")
+        MutableSet[T]
+
+        class V(MutableSet[T]):
+            pass
+        """
+        )
+        inferred = next(node.infer())
+        mro_entries = list(inferred.mro())
+        self.assertIsInstance(mro_entries[0], astroid.ClassDef)
+        self.assertEqual(mro_entries[0].name, "V")
+        self.assertIsInstance(mro_entries[1], astroid.ClassDef)
+        self.assertEqual(mro_entries[1].name, "MutableSet")
+        self.assertIsInstance(mro_entries[2], astroid.ClassDef)
+        self.assertEqual(mro_entries[2].name, "Set")
+        self.assertIsInstance(mro_entries[3], astroid.ClassDef)
+        self.assertEqual(mro_entries[3].name, "Collection")
+        self.assertIsInstance(mro_entries[4], astroid.ClassDef)
+        self.assertEqual(mro_entries[4].name, "Sized")
+        self.assertIsInstance(mro_entries[5], astroid.ClassDef)
+        self.assertEqual(mro_entries[5].name, "Iterable")
+        self.assertIsInstance(mro_entries[6], astroid.ClassDef)
+        self.assertEqual(mro_entries[6].name, "Container")
+        self.assertIsInstance(mro_entries[7], astroid.ClassDef)
+        self.assertEqual(mro_entries[7].name, "object")
 
 
 class ReBrainTest(unittest.TestCase):

--- a/tests/unittest_brain.py
+++ b/tests/unittest_brain.py
@@ -99,6 +99,11 @@ import astroid
 import astroid.test_utils as test_utils
 
 
+def assertEqualMro(klass, expected_mro):
+    """Check mro names."""
+    assert [member.name for member in klass.mro()] == expected_mro
+
+
 class HashlibTest(unittest.TestCase):
     def _assert_hashlib_class(self, class_obj):
         self.assertIn("update", class_obj)
@@ -1212,6 +1217,22 @@ class TypingBrain(unittest.TestCase):
         Test that the type aliased thanks to typing._alias function are
         correctly inferred.
         """
+
+        def check_metaclass(node: nodes.ClassDef):
+            meta = node.metaclass()
+            assert (
+                isinstance(meta, nodes.ClassDef)
+                and meta.name == "ABCMeta_typing"
+                and "ABCMeta" == meta.basenames[0]
+                and meta.locals.get("__getitem__") is not None
+            )
+            abc_meta = next(meta.bases[0].infer())
+            assert (
+                isinstance(abc_meta, nodes.ClassDef)
+                and abc_meta.name == "ABCMeta"
+                and abc_meta.locals.get("__getitem__") is None
+            )
+
         node = builder.extract_node(
             """
         from typing import TypeVar, MutableSet
@@ -1219,28 +1240,62 @@ class TypingBrain(unittest.TestCase):
         T = TypeVar("T")
         MutableSet[T]
 
-        class V(MutableSet[T]):
+        class Derived(MutableSet[T]):
             pass
         """
         )
         inferred = next(node.infer())
-        mro_entries = list(inferred.mro())
-        self.assertIsInstance(mro_entries[0], astroid.ClassDef)
-        self.assertEqual(mro_entries[0].name, "V")
-        self.assertIsInstance(mro_entries[1], astroid.ClassDef)
-        self.assertEqual(mro_entries[1].name, "MutableSet")
-        self.assertIsInstance(mro_entries[2], astroid.ClassDef)
-        self.assertEqual(mro_entries[2].name, "Set")
-        self.assertIsInstance(mro_entries[3], astroid.ClassDef)
-        self.assertEqual(mro_entries[3].name, "Collection")
-        self.assertIsInstance(mro_entries[4], astroid.ClassDef)
-        self.assertEqual(mro_entries[4].name, "Sized")
-        self.assertIsInstance(mro_entries[5], astroid.ClassDef)
-        self.assertEqual(mro_entries[5].name, "Iterable")
-        self.assertIsInstance(mro_entries[6], astroid.ClassDef)
-        self.assertEqual(mro_entries[6].name, "Container")
-        self.assertIsInstance(mro_entries[7], astroid.ClassDef)
-        self.assertEqual(mro_entries[7].name, "object")
+        check_metaclass(inferred)
+        assertEqualMro(
+            inferred,
+            [
+                "Derived",
+                "MutableSet",
+                "Set",
+                "Collection",
+                "Sized",
+                "Iterable",
+                "Container",
+                "object",
+            ],
+        )
+
+        node = builder.extract_node(
+            """
+        import typing
+        class Derived(typing.OrderedDict[int, str]):
+            pass
+        """
+        )
+        inferred = next(node.infer())
+        check_metaclass(inferred)
+        assertEqualMro(
+            inferred,
+            [
+                "Derived",
+                "OrderedDict",
+                "dict",
+                "object",
+            ],
+        )
+
+        node = builder.extract_node(
+            """
+        import typing
+        class Derived(typing.Pattern[str]):
+            pass
+        """
+        )
+        inferred = next(node.infer())
+        check_metaclass(inferred)
+        assertEqualMro(
+            inferred,
+            [
+                "Derived",
+                "Pattern",
+                "object",
+            ],
+        )
 
 
 class ReBrainTest(unittest.TestCase):

--- a/tests/unittest_brain.py
+++ b/tests/unittest_brain.py
@@ -36,7 +36,6 @@ import io
 import queue
 import re
 import os
-from typing import Collection, MutableSet
 
 try:
     import multiprocessing  # pylint: disable=unused-import

--- a/tests/unittest_brain.py
+++ b/tests/unittest_brain.py
@@ -1240,7 +1240,7 @@ class TypingBrain(unittest.TestCase):
         T = TypeVar("T")
         MutableSet[T]
 
-        class Derived(MutableSet[T]):
+        class Derived1(MutableSet[T]):
             pass
         """
         )
@@ -1249,7 +1249,7 @@ class TypingBrain(unittest.TestCase):
         assertEqualMro(
             inferred,
             [
-                "Derived",
+                "Derived1",
                 "MutableSet",
                 "Set",
                 "Collection",
@@ -1263,7 +1263,7 @@ class TypingBrain(unittest.TestCase):
         node = builder.extract_node(
             """
         import typing
-        class Derived(typing.OrderedDict[int, str]):
+        class Derived2(typing.OrderedDict[int, str]):
             pass
         """
         )
@@ -1272,7 +1272,7 @@ class TypingBrain(unittest.TestCase):
         assertEqualMro(
             inferred,
             [
-                "Derived",
+                "Derived2",
                 "OrderedDict",
                 "dict",
                 "object",
@@ -1282,7 +1282,7 @@ class TypingBrain(unittest.TestCase):
         node = builder.extract_node(
             """
         import typing
-        class Derived(typing.Pattern[str]):
+        class Derived3(typing.Pattern[str]):
             pass
         """
         )
@@ -1291,7 +1291,7 @@ class TypingBrain(unittest.TestCase):
         assertEqualMro(
             inferred,
             [
-                "Derived",
+                "Derived3",
                 "Pattern",
                 "object",
             ],

--- a/tests/unittest_brain.py
+++ b/tests/unittest_brain.py
@@ -1220,18 +1220,15 @@ class TypingBrain(unittest.TestCase):
 
         def check_metaclass(node: nodes.ClassDef):
             meta = node.metaclass()
-            assert (
-                isinstance(meta, nodes.ClassDef)
-                and meta.name == "ABCMeta_typing"
-                and "ABCMeta" == meta.basenames[0]
-                and meta.locals.get("__getitem__") is not None
-            )
+            assert isinstance(meta, nodes.ClassDef)
+            assert meta.name == "ABCMeta_typing"
+            assert "ABCMeta" == meta.basenames[0]
+            assert meta.locals.get("__getitem__") is not None
+
             abc_meta = next(meta.bases[0].infer())
-            assert (
-                isinstance(abc_meta, nodes.ClassDef)
-                and abc_meta.name == "ABCMeta"
-                and abc_meta.locals.get("__getitem__") is None
-            )
+            assert isinstance(abc_meta, nodes.ClassDef)
+            assert abc_meta.name == "ABCMeta"
+            assert abc_meta.locals.get("__getitem__") is None
 
         node = builder.extract_node(
             """
@@ -1250,6 +1247,7 @@ class TypingBrain(unittest.TestCase):
             inferred,
             [
                 "Derived1",
+                "MutableSet_typing",
                 "MutableSet",
                 "Set",
                 "Collection",
@@ -1273,6 +1271,7 @@ class TypingBrain(unittest.TestCase):
             inferred,
             [
                 "Derived2",
+                "OrderedDict_typing",
                 "OrderedDict",
                 "dict",
                 "object",

--- a/tests/unittest_brain.py
+++ b/tests/unittest_brain.py
@@ -1211,7 +1211,7 @@ class TypingBrain(unittest.TestCase):
         assert len(typing_module.locals["TypedDict"]) == 1
         assert inferred_base == typing_module.locals["TypedDict"][0]
 
-    @test_utils.require_version("3.7")
+    @test_utils.require_version("3.8")
     def test_typing_alias_type(self):
         """
         Test that the type aliased thanks to typing._alias function are
@@ -1293,6 +1293,31 @@ class TypingBrain(unittest.TestCase):
                 "Derived3",
                 "Pattern",
                 "object",
+            ],
+        )
+
+    @test_utils.require_version("3.8")
+    def test_typing_alias_side_effects(self):
+        """Test that typing._alias changes doesn't have unwanted consequences."""
+        node = builder.extract_node(
+            """
+        import typing
+        import collections.abc
+
+        class Derived(collections.abc.Iterator[int]):
+            pass
+        """
+        )
+        inferred = next(node.infer())
+        assert inferred.metaclass() is None  # Should this be ABCMeta?
+        assertEqualMro(
+            inferred,
+            [
+                "Derived",
+                # Should this be more?
+                # "Iterator_typing"?
+                # "Iterator",
+                # "object",
             ],
         )
 

--- a/tests/unittest_brain.py
+++ b/tests/unittest_brain.py
@@ -1206,6 +1206,7 @@ class TypingBrain(unittest.TestCase):
         assert len(typing_module.locals["TypedDict"]) == 1
         assert inferred_base == typing_module.locals["TypedDict"][0]
 
+    @test_utils.require_version("3.7")
     def test_typing_alias_type(self):
         """
         Test that the type aliased thanks to typing._alias function are

